cmd/snap-bootstrap/partition/encrypt.go

```diff
@@ -136,3 +162,63 @@ func cryptsetupClose(name string) error {
 	}
 	return nil
 }
+
+func cryptsetupAddKey(key EncryptionKey, rkey RecoveryKey, node string) error {
+	// create a named pipe to pass the recovery key
+	fpath := filepath.Join(dirs.SnapRunDir, "tmp-rkey")
+	if err := syscall.Mkfifo(fpath, 0600); err != nil {
+		return fmt.Errorf("cannot create named pipe: %v", err)
+	}
+	defer os.RemoveAll(fpath)
+
+	// add a new key to slot 1 reading the passphrase from the named pipe
+	// (explicitly choose keyslot 1 to ensure we have a predictable slot
+	// number in case we decide to kill all other slots later)
+	args := []string{
+		// add a new key
+		"luksAddKey",
+		// the encrypted device
+		node,
+		// batch processing, no password verification
+		"-q",
+		// read existing key from stdin
+		"--key-file", "-",
+		// store it in keyslot 1
+		"--key-slot", "1",
+		// the named pipe
+		fpath,
+	}
+
+	cmd := exec.Command("cryptsetup", args...)
+	cmd.Stdin = bytes.NewReader(key[:])
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// open the named pipe and write the recovery key
+	file, err := os.OpenFile(fpath, os.O_WRONLY, 0600)
+	if err != nil {
+		return fmt.Errorf("cannot open recovery key pipe: %v", err)
+	}
+	n, err := file.Write(rkey[:])
```

Comment:
This write is was not unit tested, I updated TestEncryptAddKey to test it.

Reply:
Thank you Michael.